internal/host/plugin/repository_host_catalog.go

```diff
@@ -164,3 +181,91 @@ func (r *Repository) ListCatalogs(ctx context.Context, scopeIds []string, opt ..
 	}
 	return hostCatalogs, nil
 }
+
+// getCatalog retrieves the *HostCatalog with the provided id.  If it is not found or there
+// is an problem getting it from the database an error is returned instead.
+func (r *Repository) getCatalog(ctx context.Context, id string) (*HostCatalog, error) {
+	const op = "plugin.(Repository).getCatalog"
+	c := allocHostCatalog()
+	c.PublicId = id
+	if err := r.reader.LookupByPublicId(ctx, c); err != nil {
+		return nil, errors.Wrap(ctx, err, op, errors.WithMsg(fmt.Sprintf("failed for: %s", id)))
+	}
+	return c, nil
+}
+
+// getPersistedDataForCatalog returns the persisted data for a catalog if
+// present.  c must have a valid Public Id and Scope Id set.
+func (r *Repository) getPersistedDataForCatalog(ctx context.Context, c *HostCatalog) (*plgpb.HostCatalogPersisted, error) {
+	const op = "plugin.(Repository).getPersistedDataForCatalog"
+	if c.PublicId == "" {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "empty public id")
+	}
+	if c.ScopeId == "" {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "empty scope id")
+	}
+	cSecret := allocHostCatalogSecret()
+	if err := r.reader.LookupWhere(ctx, cSecret, "catalog_id=?", c.GetPublicId()); err != nil {
+		if errors.IsNotFoundError(err) {
+			return nil, nil
+		}
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	if cSecret == nil {
+		return nil, nil
+	}
+	dbWrapper, err := r.kms.GetWrapper(ctx, c.ScopeId, kms.KeyPurposeDatabase)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op, errors.WithMsg("unable to get db wrapper"))
+	}
+	if err := cSecret.decrypt(ctx, dbWrapper); err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	per := &plgpb.HostCatalogPersisted{}
+	sec := map[string]interface{}{}
+	if err := json.Unmarshal(cSecret.GetSecret(), &sec); err != nil {
+		return nil, errors.Wrap(ctx, err, op, errors.WithMsg("unmarshaling secret json"))
+	}
+	per.Data, err = structpb.NewStruct(sec)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op, errors.WithMsg("marshaling to proto struct"))
+	}
+
+	return per, nil
+}
+
+// toPluginCatalog returns a host catalog, with it's secret if available, in the format expected
+// by the host plugin system.
+func toPluginCatalog(ctx context.Context, in *HostCatalog) (*pb.HostCatalog, error) {
+	const op = "plugin.toPluginCatalog"
+	if in == nil {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "nil storage plugin")
+	}
+	hc := &pb.HostCatalog{
+		Id:      in.GetPublicId(),
+		ScopeId: in.GetScopeId(),
+	}
+	if in.GetAttributes() != nil {
+		attrs := map[string]interface{}{}
+		if err := json.Unmarshal(in.GetAttributes(), &attrs); err != nil {
```

Comment:
Same here -- shouldn't the attributes be getting stored as proto marshalled instead of json marshalled?